File: remote/js/lib/RTRemoteClientConnection.js
/*
 pxCore Copyright 2005-2018 John Robinson

   Licensed under the Apache License, Version 2.0 (the "License");
   you may not use this file except in compliance with the License.
   You may obtain a copy of the License at

     http://www.apache.org/licenses/LICENSE-2.0

   Unless required by applicable law or agreed to in writing, software
   distributed under the License is distributed on an "AS IS" BASIS,
   WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
   See the License for the specific language governing permissions and
   limitations under the License.
*/

/**
 * The rt remote client connection
 *
 * @author      TCSCODER
 * @version     1.0
 */

const RTRemoteTCPTransport = require('./RTRemoteTCPTransport');
const RTRemoteObject = require('./RTRemoteObject');
const RTRemoteProtocol = require('./RTRemoteProtocol');

/**
 * the rt remote client connection class
 */
class RTRemoteClientConnection {
  /**
   * create new connection with protocol
   * @param {RTRemoteProtocol} protocol
   */
  constructor(protocol) {
    this.protocol = protocol;
  }

  /**
   * get rt remote object by object id
   * @param {string} objectId the object id
   * @return {RTRemoteObject} the returned rtRemote object
   */
  getProxyObject(objectId) {
    return new RTRemoteObject(this.protocol, objectId);
  }
}

/**
 * create tcp connection
 * @param {string} host the host name
 * @param {number|int} port the host port
 * @return {Promise<RTRemoteClientConnection>} the promise with connection
 */
function createTCPClientConnection(host, port) {
  const transport = new RTRemoteTCPTransport(host, port);

  // 1. create protocol, the second param is false mean protocol will open transport socket
  // connection and bind relate input/output events
  // 2. then use initialized protocol create connection
  return RTRemoteProtocol.create(transport, false).then(protocol => new RTRemoteClientConnection(protocol));
}

module.exports = {
  createTCPClientConnection,
};

File: remote/java/src/main/java/org/pxscene/rt/remote/RTRemoteClientConnection.java
package org.pxscene.rt.remote;

import org.pxscene.rt.RTException;
import org.pxscene.rt.RTObject;

/**
 * the connection class between client and remote.
 */
public class RTRemoteClientConnection {

  /**
   * the remote protocol
   */
  private RTRemoteProtocol proto;


  /**
   * the entity constructor with type.
   *
   * @param protocol the remote protocol
   */
  private RTRemoteClientConnection(RTRemoteProtocol protocol) {
    proto = protocol;
  }

  /**
   * create a TCP client connection
   *
   * @param host the host name
   * @param port the port
   * @return the RTRemoteClientConnection
   * @throws RTException if any other error occurred during operation
   */
  public static RTRemoteClientConnection createTCPClientConnection(String host, int port)
      throws RTException {
    RTRemoteTransport transport = new RTRemoteTCPTransport(host, port);
    return new RTRemoteClientConnection(new RTRemoteProtocol(transport, false));
  }

  /**
   * create new client-remote object
   *
   * @param objectId the remote object id
   * @return the rtObject
   */
  public RTObject getProxyObject(String objectId) {
    return new RTRemoteObject(proto, objectId);
  }
}
